torchtune/models/mistral/modules/transformer.py

```diff
@@ -0,0 +1,258 @@
+            dtype (torch.dtype): dtype for the caches.
+        """
+        for layer in self.layers:
+            layer.attn.kv_cache = KVCache(
+                batch_size=batch_size,
+                max_seq_len=self.max_seq_len,
+                num_heads=self.num_heads,
+                head_dim=self.head_dim,
+                dtype=dtype,
+            )
+
+        # causal_mask is used during inference to ensure we're attending
+        # to the right tokens
+        self.causal_mask = torch.tril(
+            torch.ones(self.max_seq_len, self.max_seq_len, dtype=torch.bool)
+        )
+
+    def reset_caches(self):
+        """Reset the key value caches."""
+        if self.layers[0].attn.kv_cache is None:
+            raise RuntimeError(
+                "Key value caches are not setup. Call ``setup_caches()`` first."
+            )
+
+        for layer in self.layers:
+            layer.attn.kv_cache.reset()
+
+    def forward(
+        self,
+        tokens: Tensor,
+        *,
+        input_pos: Optional[Tensor] = None,
+        mask: Optional[Tensor] = None,
+    ) -> Tensor:
+        """
+        Args:
+            tokens (Tensor): input tensor with shape [b x s]
+            input_pos (Optional[Tensor]): Optional tensor which contains the position
+                of the current token. This is only used during inference. Default is None
+
+        Note: At the very first step of inference, when the model is provided with a prompt,
+        ``input_pos`` would contain the positions of all of the tokens in the prompt
+        (eg: ``torch.arange(prompt_length)``). This is because we will need to compute the
+        KV values for each position.
+
+        Returns:
+            Tensor: output tensor with shape [b x s x d]
+
+        Raises:
+            ValueError: if causal_mask is set but input_pos is None
+
+        Notation used for tensor shapes:
+            - b: batch size
+            - s: sequence length
+            - d: embed dim
+            - m_s: max seq len
+        """
+        # input tensor of shape [b, s]
+        bsz, seq_len = tokens.shape
+
+        # shape: [b, s, d]
+        h = self.tok_embeddings(tokens)
+
+        if self.causal_mask is not None:
+            if input_pos is None:
+                raise ValueError(
+                    "Caches are setup, but the position of input token is missing"
+                )
+            if mask is not None:
+                raise ValueError(
+                    "An attention mask was set. Cannot use a non-causal mask for inference"
+                )
+            # shape: [1, input_pos_len, m_s]
+            # in most cases input_pos_len should be 1
+            mask = self.causal_mask[None, input_pos]
+
+        for layer in self.layers:
+            # shape: [b, s, d]
+            h = layer(h, mask=mask, input_pos=input_pos)
+
+        # shape: [b, s, d]
+        h = self.norm(h)
+
+        return h
+
+
+class TransformerLMWithValueHead(nn.Module):
```

Comment:
This class can just be a builder version of `TransformerLM` with appropriate choice of `output` (and flexibility of output return types), right? (Similar caveat to my above comment btw)

Reply:
yeah! hopefully having worked examples here has helped inform how we can think about the transformerdecoder refactor